sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1279,8 +1409,26 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
       (c, evPrim, evNull) => code"$evNull = true;"
     case TimestampType =>
       (c, evPrim, evNull) => code"$evPrim = (long) ${timestampToIntegerCode(c)};"
+    case DecimalType() if failOnIntegerOverflow =>
+      (c, evPrim, evNull) =>
+        code"""
+          try {
+            $evPrim = $c.toJavaBigInteger().longValueExact();
+          } catch (ArithmeticException e) {
+            throw new ArithmeticException("Casting $c to Long causes overflow");
```

Comment:
It seems `longValueExact` throws an exception even when the non-overflow case detected: https://github.com/bpupadhyaya/openjdk-8/blob/master/jdk/src/share/classes/java/math/BigDecimal.java#L3088

Reply:
Here it is converted as BigInteger instead of BigDecimal.

Reply:
oh, my bad. Thx!